Source/Core/Core/HW/WiimoteEmu/MotionPlus.cpp

```diff
@@ -189,100 +370,173 @@ bool MotionPlus::IsButtonPressed() const
-    std::copy(std::begin(init_data), std::end(init_data), data);
-    reg_data.cert_ready = 0x2;
-    return;
+    if (is_ext_connected)
+    {
+      DEBUG_LOG(WIIMOTE, "M+ initializing new extension.");
+
+      // The M+ automatically initializes an extension when attached.
+
+      // What we do here does not exactly match a real M+,
+      // but it's close enough for our emulated extensions which are not very picky.
+
+      // Disable encryption
+      {
+        constexpr u8 INIT_OFFSET = offsetof(Register, init_trigger);
+        std::array<u8, 1> enc_data = {0x55};
+        m_i2c_bus.BusWrite(ACTIVE_DEVICE_ADDR, INIT_OFFSET, int(enc_data.size()), enc_data.data());
+      }
+
+      // Read identifier
+      {
+        constexpr u8 ID_OFFSET = offsetof(Register, ext_identifier);
+        std::array<u8, 6> id_data = {};
+        m_i2c_bus.BusRead(ACTIVE_DEVICE_ADDR, ID_OFFSET, int(id_data.size()), id_data.data());
+        m_reg_data.passthrough_ext_id_0 = id_data[0];
+        m_reg_data.passthrough_ext_id_4 = id_data[4];
+        m_reg_data.passthrough_ext_id_5 = id_data[5];
+      }
+
+      // Read calibration data
+      {
+        constexpr u8 CAL_OFFSET = offsetof(Register, calibration_data);
+        m_i2c_bus.BusRead(ACTIVE_DEVICE_ADDR, CAL_OFFSET,
+                          int(m_reg_data.passthrough_ext_calib.size()),
+                          m_reg_data.passthrough_ext_calib.data());
+      }
+    }
+
+    // Update flag in register:
+    mplus_data.extension_connected = is_ext_connected;
+    Common::BitCastPtr<DataFormat>(data) = mplus_data;
   }
 
-  if (0x2 == reg_data.cert_ready)
-  {
-    constexpr std::array<u8, 6> init_data = {0x7f, 0xcf, 0xdf, 0x8b, 0x4f, 0x82};
-    std::copy(std::begin(init_data), std::end(init_data), data);
-    reg_data.cert_ready = 0x8;
+  // Only perform any of the following challenge logic if our timer is up.
+  if (m_progress_timer)
     return;
-  }
 
-  if (0x8 == reg_data.cert_ready)
+  // This is potentially any value that is less than cert_n and >= 2.
+  // A real M+ uses random values each run.
+  constexpr u8 magic[] = "DOLPHIN DOES WHAT NINTENDON'T.";
+
+  constexpr char cert_n[] =
+      "67614561104116375676885818084175632651294951727285593632649596941616763967271774525888270484"
+      "88546653264235848263182009106217734439508352645687684489830161";
+
+  constexpr char sqrt_v[] =
+      "22331959796794118515742337844101477131884013381589363004659408068948154670914705521646304758"
+      "02483462872732436570235909421331424649287229820640697259759264";
+
+  switch (m_reg_data.challenge_state)
   {
-    // A real wiimote takes about 2 seconds to reach this state:
-    reg_data.cert_ready = 0xe;
+  case ChallengeState::PreparingX:
+  {
+    MPI param_x;
+    param_x.ReadBinary(magic);
+
+    mbedtls_mpi_mul_mpi(&param_x, &param_x, &param_x);
+    mbedtls_mpi_mod_mpi(&param_x, &param_x, MPI(cert_n).Data());
```

Comment:
Kinda wondering why you didn't simply move those into `MPI` (as `multiply`/`modulo` methods or whatever), but at this point we're building APIs over APIs, so it probably doesn't matter too much either way. +/-0 depending on what others think.

Reply:
At that point I might as well rewrite our Common/bn on top of mbedtls to be more usable which I might do in another PR.